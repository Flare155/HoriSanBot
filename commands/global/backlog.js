const { SlashCommandBuilder, EmbedBuilder } = require('discord.js');
const { footerCreator } = require('../../utils/formatting/logFooterCreator.js');
const { calculateEmbedColor } = require('../../utils/formatting/calculateEmbedColor.js');
const { sendErrorMessage} = require('../../utils/formatting/errorMessageFormatter.js');
const { saveLog } = require('../../utils/saveLog.js');
const { DateTime } = require('luxon');
const User = require('../../models/User.js');

module.exports = {
    data: new SlashCommandBuilder()
        .setName('backlog')
        .setDescription('Log your immersion in the past!')
        .addStringOption(option =>
            option.setName('medium')
                .setDescription('The type of material you immersed in')
                .setRequired(true)
                .addChoices(
                    { name: 'Listening', value: 'Listening' },     // Audio
                    { name: 'Watchtime', value: 'Watchtime' },     // Audio-Visual
                    { name: 'YouTube', value: 'YouTube' },
                    { name: 'Anime', value: 'Anime' },
                    { name: 'Readtime', value: 'Readtime' },       // Reading
                    { name: 'Visual Novel', value: 'Visual Novel' },
                    { name: 'Manga', value: 'Manga' },
                ))
        .addStringOption(option =>
            option.setName('amount')
                .setDescription('Enter a time (e.g., 45m, 1h30m, 2m5s) or number of episodes (e.g., 10ep)')
                .setRequired(true)
            )
        .addStringOption(option =>
            option.setName('title')
                .setDescription('The title of the media')
                .setRequired(true)
            )
        .addStringOption(option =>
            option.setName('date')
                .setDescription('The date of the log (YYYY-MM-DD)')
                .setRequired(true)
            )
        .addStringOption(option =>
            option.setName('notes')
                .setDescription('Optional notes')
                .setRequired(false)
            )
        .addStringOption(option =>
            option.setName('episode_length')
                .setDescription('The length of each episode (e.g., 45m, 1h30m, 2m5s)')
                .setRequired(false)
            ),
    async execute(interaction) {
        try { 

            // Retrieve user inputs
            const medium = interaction.options.getString('medium');
            const input = interaction.options.getString('amount');
            const title = interaction.options.getString('title');
            const notes = interaction.options.getString('notes');
            const customEpisodeLength = interaction.options.getString('episode_length');
            const dateInput = interaction.options.getString('date'); // Retrieve date input
            const userData = await User.findOne({ userId: interaction.user.id });
            const userTimezone = userData ? userData.timezone : 'UTC';

            let unit = "";
            let episodes = 0;
            let seconds = 0;
            let count = 0;
            let unitLength = undefined;
            let totalSeconds = 0;

            // Regular expressions to match time and episode formats
            const timePattern = /^(?!.*ep)(?=.*[hms])(?:(\d+)h)?(?:(\d+)m)?(?:(\d+)s)?$/; // Matches inputs like 1h30m
            const episodePattern = /^(?!.*[hms])(\d+)ep$/; // Matches inputs like 10ep

            // Parse and validate the date input
            const parsedDate = parseDate(dateInput, userTimezone);
            if (!parsedDate) {
                return sendErrorMessage(interaction, 'Invalid date format. Please use YYYY-MM-DD.');
            }

            // Add backlog restriction here
            const maxBacklogYears = 40; // Maximum time in past allowed to backlog (40 years)
            const currentDate = DateTime.now().setZone(userTimezone);
            const yearsDifference = currentDate.diff(DateTime.fromJSDate(parsedDate), 'years').years;
            if (yearsDifference > maxBacklogYears) {
                return sendErrorMessage(interaction, `You can only backlog up to ${maxBacklogYears} years in the past.`);
            }

            // Calculate log information based on input
            if (!episodePattern.test(input) && !timePattern.test(input)) {
                return sendErrorMessage(interaction, "Invalid input format. Examples: 2ep, 1h30m, 45m. See /help log for more info.");
            }

            // Handle episodes logic
            if (episodePattern.test(input)) {
                // Ensure only Anime can be logged as Episodes
                if (medium !== "Anime") {
                    return sendErrorMessage(interaction, "You can only log Anime as Episodes. See /help log for more info.");
                }

                // Parse episodes
                episodes = parseEpisodes(input);
                if (episodes === null) {
                    return sendErrorMessage(interaction, "Invalid episodes format. Example: 10ep.");
                }

                // If the user enters a custom episode length
                if (customEpisodeLength) {
                    if (timePattern.test(customEpisodeLength)) {
                        // Parse episode length
                        unitLength = parseTime(customEpisodeLength);
                        if (unitLength === null) {
                            return sendErrorMessage(interaction, "Invalid episode length format. Examples: 45m, 1h30m.");
                        }
                    } else {
                        return sendErrorMessage(interaction, "Invalid episode length format. Examples: 45m, 1h30m.");
                    }
                } else {
                    // Default episode length is 21 minutes (1260 seconds)
                    unitLength = 1260;
                }

                unit = "Episodes";
                count = episodes;
                totalSeconds = unitLength * episodes;

            } else if (timePattern.test(input)) {
                // Handle time-based logs
                if (medium === "Anime") {
                    return sendErrorMessage(interaction, "For custom anime episode lengths, use episode_length along with episodes. See /help log for more info.");
                }

                // Parse time input
                seconds = parseTime(input);
                if (seconds === null) {
                    return sendErrorMessage(interaction, "Invalid time format. Examples: 1h30m, 2m5s.");
                }

                unit = "Seconds";
                count = seconds;
                totalSeconds = seconds;

            } else {
                // This else block is redundant due to earlier validation but kept for safety
                return sendErrorMessage(interaction, "Invalid input format. Examples: 2ep, 1h30m, 45m. See /help log for more info.");
            }

            // Error handling for invalid log amounts
            if (totalSeconds < 60) {
                return sendErrorMessage(interaction, `The minimum log size is 1 minute (60 seconds). You entered ${totalSeconds} seconds.`);
            }
            if (totalSeconds > 72000) { // 1200 minutes = 72000 seconds
                return sendErrorMessage(interaction, `The maximum log size is 1200 minutes (20 hours). You entered ${Math.round((totalSeconds / 60) * 10) / 10} minutes.`);
            }

            // Calculate title and description for embed
            const description = unit === "Episodes"
                ? `${Math.round((unitLength * 10) / 60) / 10} minutes/episode ‚Üí +${Math.round((totalSeconds * 10) / 60) / 10} points`
                : `1 point/min ‚Üí +${Math.round((totalSeconds * 10) / 60) / 10} points`;
            let embedTitle;
            if (unit !== "Episodes") {
                embedTitle = `üéâ ${interaction.user.displayName} Logged ${Math.round((totalSeconds / 60) * 10) / 10} Minutes of ${medium}!`;
            } else {
                embedTitle = `üéâ ${interaction.user.displayName} Logged ${input} of ${medium}!`;
            }

            // Check if the log is in the future
            if (DateTime.fromJSDate(parsedDate).toMillis() > currentDate.toMillis()) {
                return sendErrorMessage(interaction, "You can't log activities for future dates.");
            }

            const isBackLog = true;
            // Save the log data to the database, including the parsed date
            await saveLog(interaction, parsedDate, medium, title, notes, isBackLog, unit, count, unitLength, totalSeconds);
            // Send an embed message with the log details
            await sendLogEmbed(interaction, embedTitle, description, medium, input, totalSeconds, title, notes, parsedDate);
        } catch (error) {
            console.log(error);
            return sendErrorMessage(interaction, "An unexpected error occurred executing log command. Please try again later.")
        }
    },
};

// Utility function to create and send the embed message
async function sendLogEmbed(interaction, embedTitle, description, medium, input, totalSeconds, title, notes, date) {
    // Calculate the embed color based on the points
    const embedColor = calculateEmbedColor(totalSeconds);
    // Create footer message
    const footer = footerCreator(interaction, totalSeconds);

    const logEmbed = new EmbedBuilder()
        .setColor(embedColor)
        .setTitle(embedTitle)
        .setDescription(description)
        .setThumbnail(interaction.user.displayAvatarURL())
        .addFields(
            { name: 'üìñ Title', value: title, inline: true },
            { name: 'üìÖ Date', value: date.toISOString().split('T')[0], inline: true }, // Display date in ISO format
            { name: 'üìö Medium', value: medium, inline: true },
            { name: 'üìä Input', value: input, inline: true },
        );

    if (notes) {
        logEmbed.addFields({ name: 'üìù Notes', value: notes, inline: true });
    }
    logEmbed.setFooter(footer);

    // Send the embed
    await interaction.reply({ embeds: [logEmbed] });
}

// Utility function to parse time strings
const parseTime = (input) => {
    const timePattern = /^(?!.*ep)(?=.*[hms])(?:(\d+)h)?(?:(\d+)m)?(?:(\d+)s)?$/;
    const match = input.match(timePattern);
    if (!match) return null;
    const hours = parseInt(match[1] || 0, 10);
    const minutes = parseInt(match[2] || 0, 10);
    const seconds = parseInt(match[3] || 0, 10);
    return (hours * 3600) + (minutes * 60) + seconds;
};

// Utility function to parse episode input
const parseEpisodes = (input) => {
    const episodePattern = /^(?!.*[hms])(\d+)ep$/;
    const match = input.match(episodePattern);
    if (!match) return null;
    return parseInt(match[1], 10);
};

// Utility function to parse date strings in YYYY-MM-DD format
const parseDate = (input, userTimezone) => {
    // Regular expression to match YYYY-MM-DD
    const datePattern = /^(\d{4})-(\d{2})-(\d{2})$/;
    const match = input.match(datePattern);
    if (!match) return null;
    const year = parseInt(match[1], 10);
    const month = parseInt(match[2], 10)
    const day = parseInt(match[3], 10);

    // Create DateTime in user's timezone at local midnight
    const dt = DateTime.fromObject(
        { year, month, day, hour: 12, minute: 0, second: 0, millisecond: 0 },
        { zone: userTimezone }
    );

    // Validate the date components using Luxon
    if (
        dt.year !== year ||
        dt.month !== month ||
        dt.day !== day
    ) {
        return null;
    }
    // Convert to UTC for storage
    const utcDt = dt.toUTC();
    return utcDt.toJSDate(); // Store this Date object in UTC
};
