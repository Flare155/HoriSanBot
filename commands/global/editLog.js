const {
    SlashCommandBuilder,
    EmbedBuilder,
    ActionRowBuilder,
    ButtonBuilder,
    ButtonStyle,
} = require('discord.js');
const mongoose = require('mongoose');
const Log = require('../../models/Log');

const MAX_EPISODES = 57;
const MAX_MINUTES = 1200;

// Utility function to send error messages
async function sendErrorMessage(interaction, message) {
    await interaction.editReply(`‚ùå \`${message}\``);
}

// Utility function to parse time strings
const parseTime = (input) => {
    const timePattern = /^(?!.*ep)(?=.*[hms])(?:(\d+)h)?(?:(\d+)m)?(?:(\d+)s)?$/;
    const match = input.match(timePattern);
    if (!match) return null;
    const hours = parseInt(match[1] || 0, 10);
    const minutes = parseInt(match[2] || 0, 10);
    const seconds = parseInt(match[3] || 0, 10);
    return hours * 3600 + minutes * 60 + seconds;
};

// Utility function to parse episode input
const parseEpisodes = (input) => {
    const episodePattern = /^(?!.*[hms])(\d+)ep$/;
    const match = input.match(episodePattern);
    if (!match) return null;
    return parseInt(match[1], 10);
};

// Function to construct input string from amount object
function constructInputString(amount) {
    if (amount.unit === 'Seconds') {
        let totalSeconds = amount.count;
        let hours = Math.floor(totalSeconds / 3600);
        let minutes = Math.floor((totalSeconds % 3600) / 60);
        let seconds = totalSeconds % 60;
        let parts = [];
        if (hours > 0) parts.push(`${hours}h`);
        if (minutes > 0) parts.push(`${minutes}m`);
        if (seconds > 0) parts.push(`${seconds}s`);
        return parts.join('') || '0s';
    } else if (amount.unit === 'Episodes') {
        return `${amount.count}ep`;
    } else {
        return '';
    }
}

module.exports = {
    data: new SlashCommandBuilder()
        .setName('editlog')
        .setDescription('Edits a specific log by its ID.')
        .addStringOption((option) =>
            option
                .setName('log_id')
                .setDescription('The ID of the log to edit')
                .setRequired(true)
        )
        .addStringOption((option) =>
            option
                .setName('amount')
                .setDescription('The new amount (e.g., 1h30m, 2ep)')
                .setRequired(false)
        )
        .addStringOption((option) =>
            option
                .setName('title')
                .setDescription('The new title')
                .setRequired(false)
        )
        .addStringOption((option) =>
            option
                .setName('notes')
                .setDescription('The new notes')
                .setRequired(false)
        ),
    async execute(interaction) {
        await interaction.deferReply();

        // Get the values from the user's input
        const logId = interaction.options.getString('log_id');
        const newInput = interaction.options.getString('amount');
        const newTitle = interaction.options.getString('title');
        const newNotes = interaction.options.getString('notes');

        // Check if all optional fields are missing (null or undefined)
        if (
            newInput === null &&
            newTitle === null &&
            newNotes === null
        ) {
            await sendErrorMessage(
                interaction,
                'You must provide at least one field to update (amount, title, or notes).'
            );
            return;
        }

        // Validate logId
        if (!mongoose.Types.ObjectId.isValid(logId)) {
            await sendErrorMessage(interaction, 'Invalid log ID provided.');
            return;
        }

        try {
            // Find the log in the database
            const log = await Log.findOne({
                _id: logId,
                userId: interaction.user.id,
            });

            if (!log) {
                await sendErrorMessage(interaction, `No log found with ID \`${logId}\`.`);
                return;
            }

            // Prepare the original embed
            const originalEmbed = new EmbedBuilder()
                .setColor(0xff0000)
                .setTitle(`Original Log`)
                .setThumbnail(interaction.user.displayAvatarURL({ dynamic: true }))
                .addFields(
                    { name: 'üìö Medium', value: log.medium, inline: true },
                    { name: 'üìä Amount', value: constructInputString(log.amount), inline: true },
                    { name: '\u200B', value: '\u200B', inline: true },
                    { name: 'üìñ Title', value: log.title, inline: true }
                );

            if (log.notes) {
                originalEmbed.addFields(
                    { name: 'üìù Notes', value: log.notes, inline: true },
                    { name: '\u200B', value: '\u200B', inline: true }
                );
            }

            // Collect all new values
            let updatedInput = constructInputString(log.amount);
            let updatedTitle = log.title;
            let updatedNotes = log.notes;
            let updatedAmount = { ...log.amount };

            const identicalFields = [];

            // Function to recalculate amount
            const recalculateAmount = (medium, input) => {
                let amount = {
                    unit: '',
                    count: 0,
                    coefficient: undefined,
                    totalSeconds: 0
                };

                const time = parseTime(input);
                const episodesCount = parseEpisodes(input);

                if (medium === 'Anime') {
                    if (episodesCount !== null) {
                        if (episodesCount > MAX_EPISODES) {
                            throw new Error(
                                `The amount cannot exceed ${MAX_EPISODES} episodes for Anime.`
                            );
                        }
                        amount.unit = 'Episodes';
                        amount.count = episodesCount;
                        // If coefficient is already set, use it; otherwise, default to 1260 seconds (21 minutes) for anime
                        if (log.amount && log.amount.coefficient) {
                            amount.coefficient = log.amount.coefficient;
                        } else {
                            amount.coefficient = 1260; // 21 minutes in seconds
                        }
                        amount.totalSeconds = amount.count * amount.coefficient;
                    } else {
                        throw new Error(
                            'For the Anime medium, the amount must be in the format of episodes (e.g., 2ep).'
                        );
                    }
                } else {
                    if (episodesCount !== null) {
                        throw new Error(
                            `The amount for ${medium} must be time-based (e.g., 1h30m).`
                        );
                    }
                    if (time !== null) {
                        if (time > MAX_MINUTES * 60) {
                            throw new Error(
                                `The amount cannot exceed ${MAX_MINUTES} minutes.`
                            );
                        }
                        amount.unit = 'Seconds';
                        amount.count = time; // count is total seconds
                        amount.totalSeconds = time;
                    } else {
                        throw new Error(
                            'Invalid amount format. Examples: 2ep, 1h30m, 45m, 2m5s.'
                        );
                    }
                }

                return amount;
            };

            // Validate and update input
            if (newInput) {
                const existingInput = constructInputString(log.amount);
                if (newInput === existingInput) {
                    identicalFields.push('amount');
                } else {
                    try {
                        const amount = recalculateAmount(log.medium, newInput);
                        updatedInput = newInput;
                        updatedAmount = amount;
                    } catch (err) {
                        await sendErrorMessage(interaction, `Error: ${err.message}`);
                        return;
                    }
                }
            }

            // Validate and update title
            if (newTitle) {
                if (newTitle.length > 100) {
                    await sendErrorMessage(
                        interaction,
                        'The title cannot exceed 100 characters.'
                    );
                    return;
                } else if (newTitle === log.title) {
                    identicalFields.push('title');
                } else {
                    updatedTitle = newTitle;
                }
            }

            // Validate and update notes
            if (newNotes) {
                if (newNotes.length > 1024) {
                    await sendErrorMessage(
                        interaction,
                        'The notes cannot exceed 1024 characters.'
                    );
                    return;
                } else if (newNotes === log.notes) {
                    identicalFields.push('notes');
                } else {
                    updatedNotes = newNotes;
                }
            }

            if (identicalFields.length === 3) { // Only input, title, notes are tracked
                await sendErrorMessage(
                    interaction,
                    'All provided fields are identical to the original. No changes made.'
                );
                return;
            } else if (identicalFields.length > 0) {
                const fieldsList = identicalFields.map(field => `\`${field}\``).join(', ');
                await interaction.followUp(
                    `‚ö†Ô∏è The following fields are identical to the original and were not updated: ${fieldsList}.`
                );
            }

            // Update the log object with new values
            log.title = updatedTitle;
            log.notes = updatedNotes;
            log.amount = updatedAmount;

            // Create an updated embed after the changes
            const updatedEmbed = new EmbedBuilder()
                .setColor(0x00ff00)
                .setTitle('Updated Log')
                .setThumbnail(
                    interaction.user.displayAvatarURL({ dynamic: true })
                )
                .addFields(
                    { name: 'üìö Medium', value: log.medium, inline: true },
                    { name: 'üìä Amount', value: constructInputString(updatedAmount), inline: true },
                    { name: '\u200B', value: '\u200B', inline: true },
                    { name: 'üìñ Title', value: updatedTitle, inline: true }
                );

            if (updatedNotes) {
                updatedEmbed.addFields(
                    { name: 'üìù Notes', value: updatedNotes, inline: true },
                    { name: '\u200B', value: '\u200B', inline: true }
                );
            }

            // Create buttons for confirmation
            const row = new ActionRowBuilder().addComponents(
                new ButtonBuilder()
                    .setCustomId('confirm_edit')
                    .setLabel('Confirm Changes')
                    .setStyle(ButtonStyle.Success),
                new ButtonBuilder()
                    .setCustomId('cancel_edit')
                    .setLabel('Cancel')
                    .setStyle(ButtonStyle.Secondary)
            );

            // Send both embeds in one message
            await interaction.editReply({
                embeds: [originalEmbed, updatedEmbed],
                components: [row],
            });

            // Create a collector to handle button interactions
            const filter = (i) => i.user.id === interaction.user.id;
            const collector = interaction.channel.createMessageComponentCollector({
                filter,
                time: 15000,
            });

            collector.on('collect', async (i) => {
                if (i.customId === 'confirm_edit') {
                    // User confirmed the edit
                    await log.save();
                    await i.update({
                        content: `‚úÖ Log: **${log.title}** has been edited successfully.`,
                        components: [],
                        embeds: [],
                    });
                } else if (i.customId === 'cancel_edit') {
                    // User canceled the edit
                    await i.update({
                        content:
                            '‚ùå Edit canceled. No changes were made to the log.',
                        components: [],
                        embeds: [],
                    });
                }
                collector.stop();
            });

            collector.on('end', async (collected) => {
                if (collected.size === 0) {
                    await interaction.editReply({
                        content: '‚åõ No response received. Edit canceled.',
                        components: [],
                        embeds: [],
                    });
                }
            });
        } catch (error) {
            console.error(error);
            await sendErrorMessage(interaction, 'An error occurred while updating the log.');
        }
    },
};
