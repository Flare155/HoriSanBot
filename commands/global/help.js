const { SlashCommandBuilder, EmbedBuilder } = require('discord.js');

module.exports = {
    data: new SlashCommandBuilder()
        .setName('help')
        .setDescription('Learn about the bot and overview of the commands!')
        .addStringOption(option =>
            option.setName('command')
                .setDescription('The command you need help with')
                .setRequired(false) // Optional to allow general help overview
                .addChoices(
                    { name: 'help', value: 'help' },
                    { name: 'log', value: 'log' },
                    { name: 'logs', value: 'logs' },
                    { name: 'undo', value: 'undo' },
                    { name: 'deletelog', value: 'deletelog' },
                    { name: 'profile', value: 'profile' },
                    { name: 'graph', value: 'graph' },
                    { name: 'selftimeout', value: 'selftimeout' }
                )
        ),
    async execute(interaction) {
        const command = interaction.options.getString('command');
        const helpMessages = {
            help: {
                title: 'üí° Help Command',
                description: 'Provides information about the bot and an overview of all available commands.',
                usage: '/help [command]',
                details: 'Use this command to get details on specific commands or to view the list of all commands.'
            },
            log: {
                title: 'üìù Log Command',
                description: 'Log your immersion with a medium, title, amount, and notes.',
                usage: '/log [medium] [title] [amount] [notes] [episode length=21]',
                details: `Use this command to keep track of your immersion details such as books, movies, or other media.
Use this command to keep track of your immersion details such as books, movies, or other media.
* The [amount] field can be in episodes or in raw minutes.
Examples: `5ep`(21*5=105) `1h15m` (75) `5m30s` (5.5) 
* Suggested title format: 
YouTube: `ChannelName Title`
Anime:    `SeriesName Season`
Podcast: `Creator PodcastName`
etc.
* Suggested note format:
YouTube: `[link]`
Anime:     `eps[start]-[end]`
Podcast: `Episode [number] | [link]`
-# For any additional assistance DM or @ flarenotfound on discord!
                    `
            },
            logs: {
                title: 'üìú Logs Command',
                description: 'View your log history and see IDs of logs.',
                usage: '/logs',
                details: 'Displays all your previous logs including their IDs, which can be used to delete or review entries.'
            },
            undo: {
                title: '‚Ü©Ô∏è Undo Command',
                description: 'Undo the last log you created. **Irreversible!**',
                usage: '/undo',
                details: 'This will remove your last logged entry. Be careful, as this action cannot be undone.'
            },
            deletelog: {
                title: '‚ùå Delete Log Command',
                description: 'Delete a specific log by its ID. **Irreversible!**',
                usage: '/deletelog [ID]',
                details: 'Use this command to permanently remove a log entry by its ID. Make sure to double-check before deleting!'
            },
            profile: {
                title: 'üë§ Profile Command',
                description: 'Display an overview of all your logged immersion.',
                usage: '/profile',
                details: 'Shows you an overview of your total immersion activities, including time spent on different media.'
            },
            graph: {
                title: 'üìä Graph Command',
                description: 'Visualize your immersion with a graph.',
                usage: '/graph [time period]',
                details: 'Generates a graph to visualize your immersion history over time.'
            },
            selftimeout: {
                title: '‚è±Ô∏è Self Timeout Command',
                description: 'Time yourself out so you can focus on your immersion!',
                usage: '/selftimeout [minutes]',
                details: 'Blocks you from interacting with the bot for a set amount of time to help you focus on your immersion.'
            }
        };

        if (command) {
            const helpMessage = helpMessages[command];
            const embed = new EmbedBuilder()
                .setColor('#c3e0e8')
                .setTitle(helpMessage.title)
                .setDescription(helpMessage.description)
                .addFields(
                    { name: 'Usage', value: `\`${helpMessage.usage}\`` },
                    { name: 'Details', value: helpMessage.details } 
                    // Can add any sort of details or additional information about each option inside the command here
                )
                .setFooter({ text: 'For any additional assistance DM or @ flarenotfound on discord!' });

            await interaction.reply({ embeds: [embed] });
        } else {
            const embed = new EmbedBuilder()
                .setColor('#c3e0e8')
                .setTitle('üìú Bot Commands Overview')
                .setDescription('Here is a list of commands you can use with this bot:\nUse **/help [command]** for detailed info on a specific command.\n‚Äã')
                .addFields(
                    { name: 'üí° `/help`', value: 'Learn about the bot and overview of the commands.', inline: false },
                    { name: 'üìù `/log`', value: 'Log your immersion with a medium, title, amount, and notes.', inline: false },
                    { name: 'üìú `/logs`', value: 'View your log history and see IDs of logs.', inline: false },
                    { name: '‚Ü©Ô∏è `/undo`', value: 'Undo the last log you created. **Irreversible!**', inline: false },
                    { name: '‚ùå `/deletelog`', value: 'Delete a specific log by its ID. **Irreversible!**', inline: false },
                    { name: 'üë§ `/profile`', value: 'Display an overview of all your logged immersion', inline: false },
                    { name: 'üìä `/graph`', value: 'Visualize your immersion with a graph.', inline: false },
                    { name: '‚è±Ô∏è `/selftimeout`', value: 'Time yourself out so you can focus on your immersion!', inline: false }
                )
                .setFooter({ text: 'For any additional assistance DM or @ flarenotfound on discord!' });

            await interaction.reply({ embeds: [embed] });
        }
    },
};
