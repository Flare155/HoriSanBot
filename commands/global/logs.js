const fs = require('fs');
const { AttachmentBuilder, SlashCommandBuilder, EmbedBuilder } = require('discord.js');
const Log = require('../../models/Log');
const User = require('../../models/User');  // Import the User model
const { localTimeConverter } = require('../../utils/localTimeConverter'); // Import streak utility

module.exports = {
  data: new SlashCommandBuilder()
    .setName('logs')
    .setDescription('Sends your log history!')
    .addStringOption(option =>
      option.setName('medium')
        .setDescription('Medium for filtering')
        .setRequired(true)
        .addChoices(
          { name: 'All', value: 'All' },
          // Audio
          { name: 'Listening', value: 'Listening' },
          // Audio-Visual
          { name: 'Watchtime', value: 'Watchtime' },
          { name: 'YouTube', value: 'YouTube' },
          { name: 'Anime', value: 'Anime' },
          // Reading
          { name: 'Readtime', value: 'Readtime' },
          { name: 'Visual Novel', value: 'Visual Novel' },
          { name: 'Manga', value: 'Manga' },
        ))
    .addUserOption(option =>
      option.setName('user')
        .setDescription('Optionally input a user to see other peoples logs')
        .setRequired(false)
    ),

  async execute(interaction) {
    await interaction.deferReply();
    try {
      const medium = interaction.options.getString('medium');
      const user = interaction.options.getUser('user') || interaction.user;
      const userId = user.id;
      
      // Fetch the user's timezone from the database
      const userData = await User.findOne({ userId: userId});
      const userTimezone = userData ? userData.timezone : 'UTC';

      // Fetch logs based on the medium filter
      let logs;
      if (medium == "All") {
        logs = await Log.find({
          userId: userId,
        });
      } else {
        logs = await Log.find({
          userId: userId,
          medium: medium
        });
      }

      if (!logs || logs.length === 0) {
        return interaction.editReply(`No logs of medium ${medium} found`);
      }

      // Sort logs by timestamp in descending order (newest first)
      logs.sort((a, b) => new Date(b.timestamp) - new Date(a.timestamp));

      // Slice to get the first three logs
      const firstThreeLogs = logs.slice(0, 3);

      // Display the first 3 logs in a detailed embed
      const embed = new EmbedBuilder()
        .setColor('#c3e0e8')
        .setTitle(`${user.displayName}'s Recent Logs (${medium})`)
        .setThumbnail(user.displayAvatarURL({ dynamic: true }))
        .setFooter({ text: 'Use /undo or /deletelog to remove a log' });

      // Iterate over the first three logs
      firstThreeLogs.forEach((log) => {
        // Format time using the user's timezone, without seconds
        const formattedDate = localTimeConverter(log.timestamp, userTimezone);
        const formattedAmount = `${log.amount.totalSeconds} Seconds`; // Needs updating
        const title = log.title || 'N/A';
        const notes = log.notes || 'N/A';
        const isBacklog = log.isBackLog || 'N/A';

        // If I put these tabbed in on mobile looks like crap
        embed.addFields({
          name: `\`${formattedDate}\``,
          value: `**Medium**: ${log.medium}
**Amount**: ${formattedAmount}
**Title**: ${title}
**Notes**: ${notes !== 'N/A' ? notes : 'No notes provided'}
**ID**: ${log._id}
**Backlog**: ${isBacklog}\n`
        });
      });

      // Send the embed
      await interaction.editReply({ embeds: [embed] });

      // Create and send the full log history as a text file
      const formattedLogs = logs.map((log) => {
        const formattedDate = localTimeConverter(log.timestamp, userTimezone);
        let formattedAmount = `${log.amount.totalSeconds} Seconds`; // Needs updating
        let title = log.title || 'N/A';
        let notes = log.notes || 'N/A';
        let isBacklog = log.isBackLog || 'N/A';

        // Remove unicode range u1F000-1FFFF (emoji)
        // Remove animated <a:name:id> / regular emojis <name:id>
        // Remove discord emojis :name:
        var notes_no_emojis = notes.replace(/[\u{1F000}-\u{1FFFF}]|<a?:\w+:\d+>|:.+?:/gmu, ""); 
        var title_no_emojis = title.replace(/[\u{1F000}-\u{1FFFF}]|<a?:\w+:\d+>|:.+?:/gmu, ""); 

        let logString = `⏤⏤⏤⏤⏤⏤\n${formattedDate}\nMedium: ${log.medium} (${formattedAmount})`;
        if (title != 'N/A') {
          logString += `\nTitle: ${title_no_emojis}`;
        };
        if (notes != 'N/A') {
          logString += `\nNotes: ${notes_no_emojis}`;
        };
        if (isBacklog != 'N/A') {
          logString += `\nBacklog: ${isBacklog}`;
        }
        logString += `\nID: ${log._id}`;
        return logString;
      });

      const logText = formattedLogs.join('\n');
      fs.writeFileSync('logs.txt', logText);
      const attachment = new AttachmentBuilder('./logs.txt');
      interaction.followUp({ files: [attachment] });

    } catch (error) {
      console.error(error);
      interaction.editReply('Error fetching logs');
    }
  },
};
