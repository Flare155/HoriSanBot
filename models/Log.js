const mongoose = require('mongoose');

const logSchema = new mongoose.Schema({
    userId: { type: String, required: true },
    guildId: { type: String, required: true },
    timestamp: { type: Date, required: true },
    medium: { type: String, required: true},
    unit: { type: String, required: true},
    amount: { type: Number, required: true, default: 0, min: 0 },
    points: { type: Number, required: true, default: 0, min: 0 },
    title: { type: String, required: true},
    notes: { type: String, required: false},
});

logSchema.index({ userId: 1 });
logSchema.index({ timestamp: -1 });

const Log = mongoose.model('Log', logSchema);
module.exports = Log;


"SELECT "