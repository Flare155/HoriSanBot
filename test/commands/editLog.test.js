import { expect } from 'vitest';
import { execute } from '../../commands/global/editLog';
import Log from '../../models/Log';
import {myTest, createMockInteraction} from '../fixtures';

myTest('should return error when log not found by id', async ({ interaction }) => {
    await execute(interaction);
    expect(interaction.editReply).toMatchSnapshot();
});

myTest('should edit log', async ({ log, interaction }) => {
    let editInteraction = createMockInteraction({
        log_id: log.id,
        title: 'new title'
    });
    editInteraction.user = interaction.user;
    editInteraction.guild = interaction.guild;

    await execute(editInteraction);
    await confirmEdit(editInteraction);
    const editedLog = await Log.findById(log._id);
    expect(editedLog.title).toBe('new title');
});

myTest('should cancel edit log', async ({log, interaction}) => {
    let editInteraction = createMockInteraction({
        log_id: log.id,
        title: 'new title'
    });
    editInteraction.user = interaction.user;
    editInteraction.guild = interaction.guild;

    await execute(editInteraction);
    await cancelEdit(editInteraction);
    const editedLog = await Log.findById(log._id);
    expect(editedLog.title).toBe(log.title);
});

async function confirmEdit(interaction) {
    const collector = interaction.channel.createMessageComponentCollector();
    const collectorHandler = collector.on.mock.calls[0][1];
    await collectorHandler({ customId: 'confirm_edit', update: vi.fn() });
}

async function cancelEdit(interaction) {
    const collector = interaction.channel.createMessageComponentCollector();
    const collectorHandler = collector.on.mock.calls[0][1];
    await collectorHandler({ customId: 'cancel_edit', update: vi.fn() });
}
