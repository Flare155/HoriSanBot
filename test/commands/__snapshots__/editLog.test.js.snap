// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should return error when log not found by id 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "‚ùå \`Invalid log ID provided.\`",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
